File: algorithms/0543-diameter-of-binary-tree/Solution.java
// Diameter of Binary Tree - DFS returns height; track max left+right path.
// Time O(n), Space O(h).

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int val) { this.val = val; }
}

class Solution {
    private int best = 0;

    public int diameterOfBinaryTree(TreeNode root) {
        best = 0;
        height(root);
        return best;
    }

    private int height(TreeNode node) {
        if (node == null) {
            return 0;
        }
        int left = height(node.left);
        int right = height(node.right);
        best = Math.max(best, left + right);
        return 1 + Math.max(left, right);
    }
}

File: algorithms/0543-diameter-of-binary-tree/Solution.kt
// Diameter of Binary Tree - DFS returns height; track max left+right path.
// Time O(n), Space O(h).

class TreeNode(var `val`: Int, var left: TreeNode? = null, var right: TreeNode? = null)

class Solution {
    private var best = 0

    fun diameterOfBinaryTree(root: TreeNode?): Int {
        best = 0
        height(root)
        return best
    }

    private fun height(node: TreeNode?): Int {
        if (node == null) return 0
        val left = height(node.left)
        val right = height(node.right)
        best = maxOf(best, left + right)
        return 1 + maxOf(left, right)
    }
}
